## 1. РАБОТЫ С ДАННЫМИ В PANDAS / EASY
<a class="anchor" id="0"></a>
### Cодержание:
>* [1. Предобработка данных](#1)
>* [2. Базовые операции со столбцами](#2)
>* [3. Работа с датами в `DataFrame`](#3)
>* [3.1 Формат `DATETIME`](#3-1)
>* [3.2 Выделение атрибутов `DATETIME`](#3-2)
>* [3.3 Работа с интервалами](#3-3)
>* [3.4 Метод `.diff()`](#3-4)
>* [4. Cоздание и преобразование столбцов с помощью функций](#4)
>* [5. Тип данных Category](#5)
>* [6. Практика по Pandas](#6)


Предобработка данных:
- очистка данных от аномальных значений (выбросов);
- работа с пропущенными значениями;
- удаление признаков, которые не несут полезной информации;
- создание новых признаков;
- преобразование признаков и приведение данных к необходимому для анализа и модели формату.

### FEATURE ENGINEERING / ГЕНЕРАЦИЯ ПРИЗНАКОВ
Цели:
- Научиться создавать новые признаки в данных с помощью базовых операций со столбцами.
- Освоить методы работы с датой и временем в Pandas.
- Научиться применять собственные функции для преобразования столбцов и создания новых признаков.
- Познакомиться с новым типом данных category и научиться использовать его при работе с данными.



In [9]:
import pandas as pd

melb_data = pd.read_csv('melb_data_ps.csv', sep=',')

## 2. Базовые операции со столбцами DataFrame] <a class="anchor" id="2"></a>

[К содержанию](#0)

### СОЗДАНИЕ КОПИИ ТАБЛИЦЫ
Создадим копию исходной таблицы с помощью метода `copy()`

In [7]:
melb_df = melb_data.copy()
melb_df.head()

index      Suburb           Address  Rooms Type      Price Method SellerG  \
0      0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1      1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2      2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3      3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4      4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5  ...    1     202.0         126.0       1970   
1  4/02/2016       2.5  ...    0     156.0          79.0       1900   
2  4/03/2017       2.5  ...    0     134.0         150.0       1900   
3  4/03/2017       2.5  ...    1      94.0         126.0       1970   
4  4/06/2016       2.5  ...    2     120.0         142.0       2014   

   CouncilArea  Lattitude  Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996    144.9984  Northern Metropolitan           4019   
1        Yarra   -37.8079    144.9934  Northern Metropolitan           4019   
2        Yarra   -37.8093    144.9944  Northern Metropolitan           4019   
3        Yarra   -37.7969    144.9969  Northern Metropolitan           4019   
4        Yarra   -37.8072    144.9941  Northern Metropolitan           4019   

          Coordinates  
0  -37.7996, 144.9984  
1  -37.8079, 144.9934  
2  -37.8093, 144.9944  
3  -37.7969, 144.9969  
4  -37.8072, 144.9941  

[5 rows x 23 columns]

### УДАЛЕНИЕ СТОЛБЦОВ
За удаление строк и столбцов в таблице отвечает метод `drop()`.

- `labels` — порядковые номера или имена столбцов, которые подлежат удалению; если их несколько, то передаётся список;
- `axis` — ось совершения операции, `axis=0` — удаляются строки, `axis=1` — удаляются столбцы;
- `inplace` — если параметр выставлен на `True`, происходит замена изначального DataFrame на новый, при этом метод ничего не возвращает; если на `False` — возвращается копия `DataFrame`, из которой удалены указанные строки (столбцы), при этом первоначальный `DataFrame` не изменяется; по умолчанию параметр равен `False`.

Удалим столбцы *index* и *Coordinates* из таблицы с помощью метода `drop()`.

In [3]:
melb_df = melb_df.drop(['index', 'Coordinates'], axis=1)
melb_df.head()

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5      3067  ...         1    1     202.0         126.0   
1  4/02/2016       2.5      3067  ...         1    0     156.0          79.0   
2  4/03/2017       2.5      3067  ...         2    0     134.0         150.0   
3  4/03/2017       2.5      3067  ...         2    1      94.0         126.0   
4  4/06/2016       2.5      3067  ...         1    2     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0       1970        Yarra  -37.7996    144.9984  Northern Metropolitan   
1       1900        Yarra  -37.8079    144.9934  Northern Metropolitan   
2       1900        Yarra  -37.8093    144.9944  Northern Metropolitan   
3       1970        Yarra  -37.7969    144.9969  Northern Metropolitan   
4       2014        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0          4019  
1          4019  
2          4019  
3          4019  
4          4019  

[5 rows x 21 columns]

### МАТЕМАТИЧЕСКИЕ ОПЕРАЦИИ СО СТОЛБЦАМИ
Pandas поддерживает базовые математические операции между столбцами: столбцы можно складывать, вычитать, умножать, делить между собой, а также возводить в степень. С помощью таких операций мы можем создавать новые признаки или производить преобразования над старыми.
>Например, давайте создадим переменную total_rooms, в которой будем хранить общее количество комнат в здании. Для этого выполним сложение столбцов с количеством комнат, ванн и спален:

In [4]:
total_rooms = melb_df['Rooms'] + melb_df['Bedroom'] + melb_df['Bathroom']
display(total_rooms)

0         5
1         5
2         8
3         8
4         8
         ..
13575    10
13576     8
13577     8
13578     9
13579     9
Length: 13580, dtype: int64

А теперь введём признак `MeanRoomsSquare`, который соответствует средней площади одной комнаты для каждого объекта. Для этого разделим площадь здания на полученное ранее общее количество комнат:

In [5]:
melb_df['MeanRoomsSquare'] = melb_df['BuildingArea'] / total_rooms
display(melb_df['MeanRoomsSquare'])

0        25.200000
1        15.800000
2        18.750000
3        15.750000
4        17.750000
           ...    
13575    12.600000
13576    16.625000
13577    15.750000
13578    17.444444
13579    12.444444
Name: MeanRoomsSquare, Length: 13580, dtype: float64

Можно ввести ещё один интересный признак — `AreaRatio`, коэффициент соотношения площади здания (`BuildingArea`) и площади участка (`Landsize`). Для этого разницу двух площадей поделим на их сумму:

In [6]:
melb_data['Arearatio'] = (melb_data['BuildingArea'] - melb_data['Landsize']) / (melb_data['BuildingArea'] + melb_data['Landsize'])
melb_data['Arearatio'].describe()

count    13580.000000
mean        -0.277685
std          0.573462
min         -1.000000
25%         -0.661745
50%         -0.487805
75%         -0.175562
max          1.000000
Name: Arearatio, dtype: float64

Рассмотрим три случая, чтобы понять его значение:

- Если рассматриваемые площади равны, то числитель дроби зануляется и коэффициент тоже равен 0.
- Если одна из площадей начинает доминировать над другой, то коэффициент начинает расти в отрицательную сторону, если площадь участка больше площади здания, и в положительную сторону, если наоборот.
- Наконец, в предельном случае, если площадь здания равна 0, то числитель дроби равен знаменателю со знаком минус, а коэффициент равен -1, а если площадь участка равна 0, то числитель дроби равен знаменателю со знаком плюс, а коэффициент равен 1.
>Таким образом, значение в столбце `AreaRatio` служит своеобразным указателем соотношения площадей объекта недвижимости. Для пустырей — участков без строений — он будет равен -1, для домов без территории — 1, во всех остальных случаях мы можем увидеть, какая площадь больше — здания или участка.

In [7]:
print('Пустырей (участков без строений):',melb_data[melb_data['Arearatio'] == -1].shape[0])
print('Домов без придомовой территории:',melb_data[melb_data['Arearatio'] == 1].shape[0])

Пустырей (участков без строений): 17
Домов без придомовой территории: 1939


## 3. Работа с датами в DataFrame <a class="anchor" id="3"></a>

[К содержанию](#0)

### ФОРМАТ DATETIME

<a class="anchor" id="3-1"></a>
>Pandas формат `datetime`, который записывается как `YYYY-MM-DD HH: MM: SS`, то есть составляющие времени указываются в следующем порядке: год, месяц, день, час, минута, секунда.

В наших данных дата записана в виде DD/MM/YYYY, например 3/12/2017. Посмотрим на это:

In [8]:
display(melb_df['Date'])

0         3/12/2016
1         4/02/2016
2         4/03/2017
3         4/03/2017
4         4/06/2016
            ...    
13575    26/08/2017
13576    26/08/2017
13577    26/08/2017
13578    26/08/2017
13579    26/08/2017
Name: Date, Length: 13580, dtype: object

Для того чтобы преобразовывать столбцы с датами, записанными в распространённых форматах, в формат `datetime`, можно воспользоваться функцией `pandas.to_datetime()`. В нашем случае в функции нужно указать параметр `dayfirst=True`, который будет обозначать, что в первоначальном признаке первым идет день. Преобразуем столбец `Date` в формат `datetime`, передав его в эту функцию:

In [9]:
melb_df['Date'] = pd.to_datetime(melb_df['Date'], dayfirst=True)
display(melb_df['Date'])

0       2016-12-03
1       2016-02-04
2       2017-03-04
3       2017-03-04
4       2016-06-04
           ...    
13575   2017-08-26
13576   2017-08-26
13577   2017-08-26
13578   2017-08-26
13579   2017-08-26
Name: Date, Length: 13580, dtype: datetime64[ns]

### ВЫДЕЛЕНИЕ АТРИБУТОВ DATETIME 

<a class="anchor" id="3-2"></a>
[К содержанию](#0)

Тип данных `datetime` позволяет с помощью специального `аксессора dt` *(Аксессор — это атрибут столбца, хранящий переменные, которые были строковым представлением времени, а затем были изменены с помощью `pd.to_datetime()`)* выделять составляющие времени из каждого элемента столбца, такие как:
- date — дата;
- year, month, day — год, месяц, день;
- time — время;
- hour, minute, second — час, минута, секунда;
- dayofweek — номер дня недели, от 0 до 6, где 0 — понедельник, 6 — воскресенье;
- day_name — название дня недели;
- dayofyear — порядковый день года;
- quarter — квартал (интервал в три месяца).

Например, обратившись по атрибуту `dt.year` в столбце `Date`, мы можем «достать» год продажи и понять, за какой интервал времени (в годах) представлены наши данные, а также на какой год приходится наибольшее число продаж:

In [10]:
years_sold = melb_df['Date'].dt.year
print(years_sold)
print('Min year sold:', years_sold.min())
print('Max year sold:', years_sold.max())
print('Mode year sold:', years_sold.mode()[0])


0        2016
1        2016
2        2017
3        2017
4        2016
         ... 
13575    2017
13576    2017
13577    2017
13578    2017
13579    2017
Name: Date, Length: 13580, dtype: int64
Min year sold: 2016
Max year sold: 2017
Mode year sold: 2017


>В результате обращения к атрибуту `datetime melb_df['Date'].dt.year` мы получаем объект `Series`, в котором в качестве значений выступают годы продажи объектов недвижимости. Мы можем занести результат в переменную `year_sold` и далее работать с ней как с обычным столбцом `Series` — вычислять максимум, минимум и модальное значение.

Примечание. Так как модальных значений в столбце может быть несколько, метод `mode()` возвращает объект `Series`, даже если мода в данных только одна. Чтобы сохранить стилистику вывода информации о годе продажи и выводить только число, а не `Series`, мы обращаемся к результату работы метода `mode()` по индексу 0.

Теперь попробуем понять, на какие месяцы приходится пик продаж объектов недвижимости. Для этого выделим атрибут `dt.month` и на этот раз занесём результат в столбец `MonthSale`, а затем найдём относительную частоту продаж для каждого месяца от общего количества продаж — для этого используем метод `value_counts()` с параметром `normalize` (вывод в долях):

>>Как правило, результаты заносятся в таблицу (в новый столбец), когда предполагается, что столбец будет использоваться в дальнейшем. Промежуточные и тренировочные результаты будем записывать в переменные.

In [11]:
melb_df['MonthSale'] = melb_df['Date'].dt.month
melb_df['MonthSale'].value_counts(normalize=True)

5     0.149411
7     0.145950
9     0.135862
6     0.134757
8     0.114138
11    0.082032
4     0.069882
3     0.049926
12    0.044698
10    0.040574
2     0.032622
1     0.000147
Name: MonthSale, dtype: float64

Из результатов становится ясно, что наибольшее количество продаж недвижимости приходится на май, июль и сентябрь (пятый, седьмой и девятый месяцы соответственно). Месяцами застоя при этом являются месяцы — октябрь, февраль и январь (десятый, второй и первый месяцы соответственно).

### РАБОТА С ИНТЕРВАЛАМИ
<a class="anchor" id="3-3"></a>
Часто бывает такая ситуация, что необходимо вычислять интервалы между двумя временными промежутками. Например, можно вычислить, сколько дней прошло с 1 января 2016 года до момента продажи объекта. Для этого можно просто найти разницу между датами продаж и заявленной датой, представленной в формате `datetime`:

In [12]:
delta_days = melb_df['Date'] - pd.to_datetime('2016-01-01') 
display(delta_days)

0       337 days
1        34 days
2       428 days
3       428 days
4       155 days
          ...   
13575   603 days
13576   603 days
13577   603 days
13578   603 days
13579   603 days
Name: Date, Length: 13580, dtype: timedelta64[ns]

Чтобы превратить количество дней из формата интервала в формат целого числа дней, можно воспользоваться аксессором dt для формата `timedelta` и извлечь из него атрибут days:

In [13]:
display(delta_days.dt.days)

0        337
1         34
2        428
3        428
4        155
        ... 
13575    603
13576    603
13577    603
13578    603
13579    603
Name: Date, Length: 13580, dtype: int64

Рассмотрим другой пример. Давайте создадим признак возраста объекта недвижимости в годах на момент продажи. Для этого выделим из столбца с датой продажи год и вычтем из него год постройки здания. Результат оформим в виде столбца `AgeBuilding`:

In [14]:
melb_df['AgeBuilding'] = melb_df['Date'].dt.year - melb_df['YearBuilt']
display(melb_df['AgeBuilding'])

0         46
1        116
2        117
3         47
4          2
        ... 
13575     36
13576     22
13577     20
13578     97
13579     97
Name: AgeBuilding, Length: 13580, dtype: int64

>Примечание. Обратите внимание, что, так как года кодируются целым числом, в результате мы тоже получаем целочисленный столбец — тип `int64` (а не `timedelta`).

>На самом деле столбец `AgeBuilding` дублирует информацию столбца `YearBuilt`, так как, зная год постройки здания, мы автоматически знаем его возраст. Такие признаки не стоит оставлять вместе, поэтому оставим возраст здания, так как он является более наглядным, а год постройки удалим из таблицы:

In [15]:
melb_df = melb_df.drop('YearBuilt', axis=1)

In [16]:
melb_df['Date']

0       2016-12-03
1       2016-02-04
2       2017-03-04
3       2017-03-04
4       2016-06-04
           ...    
13575   2017-08-26
13576   2017-08-26
13577   2017-08-26
13578   2017-08-26
13579   2017-08-26
Name: Date, Length: 13580, dtype: datetime64[ns]

3.3 Создайте в таблице `melb_df` признак `WeekdaySale` (день недели). Найдите, сколько объектов недвижимости было продано в выходные (суббота и воскресенье), результат занесите в переменную `weekend_count`. В качестве ответа введите результат вывода переменной `weekend_count`.

In [17]:
melb_df['WeekdaySale'] = melb_df['Date'].dt.dayofweek
weekend_count = melb_df[melb_df['WeekdaySale'] == 5].shape[0] + melb_df[melb_df['WeekdaySale'] == 6].shape[0]
weekend_count



12822

### Метод `.diff()`
<a class="anchor" id="3-4"></a>

Определяет разницу между элементами `DateTime` (по умолчанию сравнивает элемент и элемент на предыдущей строке)
>`DataFrame.diff(periods=1, axis=0)`
`periodsint, default 1`
>>Periods to shift for calculating difference, accepts negative values.
`axis{0 or ‘index’, 1 or ‘columns’}, default 0`
Take difference over rows (0) or columns (1).

## 4. Создание и преобразование столбцов с помощью функций <a class="anchor" id="4"></a>

[К содержанию](#0)

Мы можем написать некоторую функцию, которая принимает на вход один элемент столбца, каким-то образом его обрабатывает и возвращает результат, после чего применить эту функцию к каждому элементу в столбце с помощью специального метода `apply()`. В результате применения этой функции будет возвращён объект `Series`, элементы которого будут представлять результат работы этой функции.
>Рассмотрим пример. В наших данных есть столбец с адресами объектов недвижимости. Проблема этого столбца в том, что в нём слишком большое количество уникальных значений: почти на каждый объект недвижимости в таблице приходится свой уникальный адрес. Убедимся в этом, вычислив количество уникальных значений в столбце с помощью метода `nunique()`:

In [18]:
print(melb_df['Address'].nunique())

13378


Если мы прогнозируем цену объекта, то такое большое количество возможных категорий может плохо сказаться на модели, которую мы бы хотели в дальнейшем построить на наших данных. Говорят, что такой признак, скорее всего, не имеет статистической значимости, потому что не позволяет разделить данные на группы, которые можно сравнить по целевому признаку.
>Из-за таких признаков зависимость между целевым признаком, который мы хотим предсказать, и признаками, на основе которых мы делаем предсказание, становится очень сложной. При этом точность моделирования при учёте такого признака может не повыситься, а даже снизиться, а производительность однозначно резко упадёт.

Обычно подобные признаки удаляют, однако можно поступить умнее: давайте извлечём из признака адреса характеристику подтипа улицы (улица, шоссе, авеню, бульвар). Для этого сначала внимательнее посмотрим на структуру адреса, выберем несколько строк столбца Address:

In [19]:
print(melb_df['Address'].loc[177])
print(melb_df['Address'].loc[1812])
print(melb_df['Address'].loc[9001])

2/119 Railway St N
9/400 Dandenong Rd
172 Danks St


Итак, адрес строится следующим образом: сначала указывается номер дома и корпус, после указывается название улицы, а в конце — подтип улицы, но в некоторых случаях к подтипу добавляется географическая отметка (N — север, S — юг и т. д.), она нам не нужна . Для того чтобы выделить подтип улицы, на которой находится объект, можно использовать следующую функцию:

In [20]:
# На вход данной функции поступает строка с адресом.
def get_street_type(address):
# Создаём список географических пометок exclude_list.
    exclude_list = ['N', 'S', 'W', 'E']
# Метод split() разбивает строку на слова по пробелу.
# В результате получаем список слов в строке и заносим его в переменную address_list.
    address_list = address.split(' ')
# Обрезаем список, оставляя в нём только последний элемент,
# потенциальный подтип улицы, и заносим в переменную street_type.
    street_type = address_list[-1]
# Делаем проверку на то, что полученный подтип является географической пометкой.
# Для этого проверяем его на наличие в списке exclude_list.
    if street_type in exclude_list:
# Если переменная street_type является географической пометкой,
# переопределяем её на второй элемент с конца списка address_list.
        street_type = address_list[-2]
# Возвращаем переменную street_type, в которой хранится подтип улицы.
    return street_type

Теперь применим эту функцию к столбцу c адресом. Для этого передадим функцию `get_street_type` в аргумент метода столбца `apply()`. В результате получим объект `Series`, который положим в переменную `street_types`:

In [22]:
street_types = melb_df['Address'].apply(get_street_type)
display(street_types)
street_types.nunique()

0        St
1        St
2        St
3        La
4        St
         ..
13575    Cr
13576    Dr
13577    St
13578    St
13579    St
Name: Address, Length: 13580, dtype: object

56

### Обратите внимание, что функция пишется для одного элемента столбца, а метод `apply()` применяется к каждому его элементу. Используемая функция обязательно должна иметь возвращаемое значение.

In [23]:
street_types.value_counts()

St           8012
Rd           2825
Ct            612
Dr            447
Av            321
Gr            311
Pde           211
Pl            169
Cr            152
Cl            100
La             67
Bvd            53
Tce            47
Wy             40
Avenue         40
Cct            25
Hwy            24
Parade         15
Boulevard      13
Sq             11
Crescent        9
Cir             7
Strand          7
Esplanade       6
Grove           5
Gdns            4
Grn             4
Fairway         4
Mews            4
Crossway        3
Righi           3
Victoria        2
Ridge           2
Crofts          2
Esp             2
Glade           1
Gra             1
Ave             1
Woodland        1
Outlook         1
Hts             1
Highway         1
Athol           1
Summit          1
Grand           1
Res             1
Nook            1
Eyrie           1
Dell            1
East            1
Loop            1
Grange          1
Terrace         1
Cove            1
Qy              1
Corso     

Из данного вывода можно увидеть, что есть группа наиболее популярных подтипов улиц, а дальше частота подтипов быстро падает.

В таком случае давайте применим очень распространённый метод уменьшения количества уникальных категорий — выделим n подтипов, которые встречаются чаще всего, а остальные обозначим как `'other' `(другие).

Для этого к результату метода `value_counts` применим метод `nlargest()`, который возвращает n наибольших значений из `Series`. Зададим n=10, т. е. мы хотим отобрать десять наиболее популярных подтипов. Извлечём их названия с помощью атрибута `index`, а результат занесём в переменную `popular_stypes`:

In [24]:
popular_stypes = street_types.value_counts().nlargest(10).index
print(popular_stypes)

Index(['St', 'Rd', 'Ct', 'Dr', 'Av', 'Gr', 'Pde', 'Pl', 'Cr', 'Cl'], dtype='object')


Теперь, когда у нас есть список наиболее популярных подтипов улиц, введём `lambda`-функцию, которая будет проверять, есть ли строка x в этом перечне, и, если это так, `lambda`-функция будет возвращать x, в противном случае она будет возвращать строку `'other'`. Наконец, применим такую функцию к `Series` `street_types`, полученной ранее, а результат определим в новый столбец таблицы `StreetType`:

In [26]:
melb_data['StreetType'] = street_types.apply(lambda x: x if x in popular_stypes else 'other')
display(melb_data['StreetType'])
melb_data['StreetType'].nunique()

0           St
1           St
2           St
3        other
4           St
         ...  
13575       Cr
13576       Dr
13577       St
13578       St
13579       St
Name: StreetType, Length: 13580, dtype: object

11

Теперь, у нас нет потребности хранить признак Address, так как, если конкретное местоположение объекта всё же и влияет на его стоимость, то оно определяется столбцами `Longitude` и `Lattitude`. Удалим его из нашей таблицы:

In [27]:
melb_df = melb_df.drop('Address', axis=1)

Таким образом, с помощью написания собственных функций и их комбинирования с методом `apply()` из библиотеки Pandas мы смогли извлечь информацию из признака с адресом и заменить на признак подтипа улицы.

### Меньше уникальных значений в признаках!
1. Определите (хотя бы на глаз) соотношение числа уникальных категорий интересующего вас признака к общему числу объектов в таблице. Если это соотношение превышает значение 30 %, то это уже повод задуматься над уменьшением числа категорий и перейти к шагу 2.
2. Если ваш признак уникален для каждого объекта, например адрес, имя или название, то такой признак, скорее всего, не имеет статистической значимости. От таких признаков чаще всего избавляются. Однако можно попробовать выделить из этого признака какие-то общие черты, например, как мы это сделали с подтипами улиц. Такой же трюк можно произвести, например, с названиями компаний, в которых может быть скрыт признак типа организации (из строки «ООО Три Слепые Мыши» можно извлечь ООО — общество с ограниченной ответственностью).
3. Если даже после преобразования число уникальных категорий всё ещё велико, можно попробовать с помощью метода value_counts() оценить, есть ли в данных категории, которые употребляются гораздо реже, чем остальные. Если такие категории присутствуют, переходите к шагу 4.
4. Можно подобрать число  популярных категорий таким образом, чтобы эти категории покрывали большую часть ваших данных.
Когда вы выбрали оптимальное число, переходите к шагу 5.
5. Наконец, можно совершить преобразование, обозначив категории, не попавшие в число популярных, как «другие».


Документация по методу `apply()` https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.apply.html

## 5. Тип данных Category <a class="anchor" id="5"></a>

[К содержанию](#0)

>Важной частью предобработки данных является оптимизация хранения данных и работы с ними. В этом разделе мы поговорим об оптимизации работы с категориальными признаками.

### ПРИЗНАКИ: КАТЕГОРИАЛЬНЫЕ И ЧИСЛОВЫЕ

Под **числовыми** признаками обычно подразумевают признаки, которые отражают количественную меру и могут принимать значения из неограниченного диапазона.

Числовые признаки могут быть:
- **дискретными** (количество комнат, дней всё то что принимает строгие значения)
- **непрерывными** (масса, цена, площадь всё что принимает значения в диапазанах)

>Дискретные признаки чаще всего представлены целыми числами, а непрерывные — целыми числами и числами с плавающей точкой.

Под **категориальными** признаками обычно подразумевают столбцы в таблице, которые обозначают принадлежность объекта к какому-то классу/категории.

Категориальные признаки могут быть:
- номинальными (например, пол, национальность, район);
- порядковыми (например, уровень образования, уровень комфорта, стадия заболевания).

>Такие признаки имеют ограниченный набор значений. Они чаще всего представлены в виде текстового описания и кодируются в Pandas типом данных `object`.

Не существует жёсткого правила, определяющего, сколько значений должна иметь категориальная переменная. Вы должны использовать собственные знания о предметной области, чтобы сделать выбор. Однако в этом разделе мы всё же введём некоторые рекомендации.

### На самом деле классификация признаков уходит даже глубже:
https://medium.com/nuances-of-programming/статистические-типы-данных-используемые-в-машинном-обучении-b8891039b09

### ПОЧЕМУ ТАК ВАЖНО ОТЛИЧАТЬ КАТЕГОРИАЛЬНЫЕ СТОЛБЦЫ ОТ ДРУГИХ?

>Оказывается, анализ и предобработка категориальных признаков отличается от предобработки числовых признаков.

Мы уже видели, что в столбцах с типом данных `object` мы не можем рассчитать среднее значение, стандартное отклонение или другие статистические параметры, которые годятся только для чисел.

Способ заполнения пропущенных данных и поиск аномальных значений также различаются для разных типов признаков, но этот момент мы рассмотрим в модуле по очистке данных.

In [28]:
# Давайте определим число уникальных категорий в каждом столбце нашей таблицы melb_df. 
# Для этого создадим вспомогательную таблицу unique_counts:
# создаём пустой список
unique_list = []
# пробегаемся по именам столбцов в таблице
for col in melb_df.columns:
    # создаём кортеж (имя столбца, число уникальных значений)
    item = (col, melb_df[col].nunique(),melb_df[col].dtype) 
    # добавляем кортеж в список
    unique_list.append(item) 
# создаём вспомогательную таблицу и сортируем её
unique_counts = pd.DataFrame(
    unique_list,
    columns=['Column_Name', 'Num_Unique', 'Type']
).sort_values(by='Num_Unique',  ignore_index=True)
# выводим её на экран
display(unique_counts)

Column_Name  Num_Unique            Type
0              Type           3          object
1       WeekdaySale           5           int64
2            Method           5          object
3        Regionname           8          object
4             Rooms           9           int64
5          Bathroom           9           int64
6               Car          11           int64
7           Bedroom          12           int64
8         MonthSale          12           int64
9       CouncilArea          33          object
10             Date          58  datetime64[ns]
11      AgeBuilding         152           int64
12         Postcode         198           int64
13         Distance         202         float64
14          SellerG         268          object
15    Propertycount         311           int64
16           Suburb         314          object
17     BuildingArea         602         float64
18  MeanRoomsSquare        1277         float64
19         Landsize        1448         float64
20            Price        2204         float64
21        Lattitude        6503         float64
22       Longtitude        7063         float64

Если присмотреться внимательно, можно увидеть резкий скачок количества уникальных значений, начиная с 14 строки таблицы, где число уникальных значений составляет 152. Учтём этот момент.

>Условимся, что категориальными будем считать признаки, у которых число уникальных категорий меньше 150. 

Однако учтём, что признак `Date` (дата продажи), преобразованный нами ранее в формат `datetime`, является временным признаком, поэтому далее не будем его воспринимать как категориальный. 

>К тому же в наш потенциальный список попали количественные столбцы `Rooms`, `Car`, `Bedroom` и `Bathroom`. Договоримся, что мы не будем относить их к разряду категориальных, однако, как упоминалось ранее, такое тоже вполне возможно.

Примечание. Ещё раз подчеркиваем, что такая классификация признаков является исключительно субъективной и специфична для задачи.

### ТИП ДАННЫХ CATEGORY
#### Для хранения и оптимизации работы с категориальными признаками в Pandas предусмотрен специальный тип данных — `category`.

- Этот тип данных является гибридным: внешне он выглядит как строка, но внутренне представлен массивом целых чисел. Так как данные вместо изначальных строк хранятся в памяти как число, то объём памяти, занимаемой таблицей при использовании типа category, резко уменьшается, что повышает эффективность хранения и работы с таблицей.

- Более того, этот тип данных расширяет возможности работы с категориальными признаками: мы можем легко преобразовывать категории, строить графики по таким данным (что сложно сделать для типа данных object). Также резко повышается производительность операций, совершаемых с такими столбцами.

>Самый простой способ преобразования столбцов к типу данных `category` — это использование уже знакомого нам метода `astype()`, в параметры которого достаточно передать строку `'category'`.

Для начала, выведем информацию о памяти, занимаемой текущей таблицей, с помощью метода `info():`

In [29]:
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Suburb           13580 non-null  object        
 1   Rooms            13580 non-null  int64         
 2   Type             13580 non-null  object        
 3   Price            13580 non-null  float64       
 4   Method           13580 non-null  object        
 5   SellerG          13580 non-null  object        
 6   Date             13580 non-null  datetime64[ns]
 7   Distance         13580 non-null  float64       
 8   Postcode         13580 non-null  int64         
 9   Bedroom          13580 non-null  int64         
 10  Bathroom         13580 non-null  int64         
 11  Car              13580 non-null  int64         
 12  Landsize         13580 non-null  float64       
 13  BuildingArea     13580 non-null  float64       
 14  CouncilArea      12211 non-null  objec

None

Сделаем преобразование столбцов к типу данных `category`:

In [30]:
cols_to_exclude = ['Date', 'Rooms', 'Bedroom', 'Bathroom', 'Car'] # список столбцов, которые мы не берём во внимание
max_unique_count = 150 # задаём максимальное число уникальных категорий
for col in melb_df.columns: # цикл по именам столбцов
    if melb_df[col].nunique() < max_unique_count and col not in cols_to_exclude: # проверяем условие
        melb_df[col] = melb_df[col].astype('category') # преобразуем тип столбца
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Suburb           13580 non-null  object        
 1   Rooms            13580 non-null  int64         
 2   Type             13580 non-null  category      
 3   Price            13580 non-null  float64       
 4   Method           13580 non-null  category      
 5   SellerG          13580 non-null  object        
 6   Date             13580 non-null  datetime64[ns]
 7   Distance         13580 non-null  float64       
 8   Postcode         13580 non-null  int64         
 9   Bedroom          13580 non-null  int64         
 10  Bathroom         13580 non-null  int64         
 11  Car              13580 non-null  int64         
 12  Landsize         13580 non-null  float64       
 13  BuildingArea     13580 non-null  float64       
 14  CouncilArea      12211 non-null  categ

None

### ПОЛУЧЕНИЕ АТРИБУТОВ CATEGORY

У типа данных `category` есть свой специальный аксесcор cat, который позволяет получать информацию о своих значениях и преобразовывать их. Например, с помощью атрибута этого аксессора `categories` мы можем получить список уникальных категорий в столбце `Regionname`:

In [31]:
print(melb_df['Regionname'].cat.categories)

Index(['Eastern Metropolitan', 'Eastern Victoria', 'Northern Metropolitan',
       'Northern Victoria', 'South-Eastern Metropolitan',
       'Southern Metropolitan', 'Western Metropolitan', 'Western Victoria'],
      dtype='object')


А теперь посмотрим, каким образом столбец кодируется в виде чисел в памяти компьютера. Для этого можно воспользоваться атрибутом `codes`:

In [32]:
display(melb_df['Regionname'].cat.codes)

0        2
1        2
2        2
3        2
4        2
        ..
13575    4
13576    6
13577    6
13578    6
13579    6
Length: 13580, dtype: int8

С помощью метода аксессора `rename_categories()` можно легко переименовать текущие значения категорий. Для этого в данный метод нужно передать словарь, ключи которого — старые имена категорий, а значения — новые.

Рассмотрим на примере: переименуем категории признака типа постройки `Type` — заменим их на полные названия (напомним, u — unit, h — house, t — townhouse).

In [33]:
melb_df['Type'] = melb_df['Type'].cat.rename_categories({
    'u': 'unit',
    't': 'townhouse',
    'h': 'house'
})
display(melb_df['Type'])

0        house
1        house
2        house
3        house
4        house
         ...  
13575    house
13576    house
13577    house
13578    house
13579    house
Name: Type, Length: 13580, dtype: category
Categories (3, object): ['house', 'townhouse', 'unit']

Подробнее ознакомиться с другими возможностями типа данных `category`:
https://pandas.pydata.org/pandas-docs/stable/user_guide/categorical.html

### ПОДВОДНЫЕ КАМНИ

А теперь представим ситуацию, что появилась новая партия домов и теперь мы продаём и квартиры (`flat`). Создадим объект `Series new_houses_types`, в котором будем хранить типы зданий новой партии домов. Преобразуем тип `new_houses_types` в такой же тип, как и у столбца Type в таблице `melb_data`, и выведем результат на экран:

In [34]:
new_houses_types = pd.Series(['unit', 'house', 'flat', 'flat', 'house'])
new_houses_types = new_houses_types.astype(melb_df['Type'].dtype)
display(new_houses_types)

0     unit
1    house
2      NaN
3      NaN
4    house
dtype: category
Categories (3, object): ['house', 'townhouse', 'unit']

На самом деле причина проста: тип данных `category` хранит только категории, которые были объявлены при его инициализации. При встрече с новой, неизвестной ранее категорией, этот тип превратит её в пустое значение, так как он просто не знает о существовании этой категории.

Решить эту проблему на самом деле не сложно. Можно добавить категорию `flat` в столбец `Type` с помощью метода акссесора `cat add_categories()`, в который достаточно просто передать имя новой категории:

In [35]:
melb_df['Type'] = melb_df['Type'].cat.add_categories('flat')
new_houses_types = pd.Series(['unit', 'house', 'flat', 'flat', 'house'])
new_houses_types = new_houses_types.astype(melb_df['Type'].dtype)
display(new_houses_types)

0     unit
1    house
2     flat
3     flat
4    house
dtype: category
Categories (4, object): ['house', 'townhouse', 'unit', 'flat']

Примечание. Добавление новой категории в столбец `Type` не отразится на самом столбце — текущие категории не изменятся, однако такое преобразование позволит добавлять в таблицу новые данные о домах с новой категорией — `flat`.
>Из данного примера можно сделать вывод, что если набор категорий в столбце жёстко не зафиксирован и может обновляться в процессе работы, то тип category не является подходящим типом данных для этого столбца или необходимо постоянно писать проверки при обновлении таблицы.

Теперь, когда мы рассмотрели основные моменты и нюансы работы с типом данных category, можно сформулировать несколько рекомендаций по его использованию:

- Необязательно каждый раз преобразовывать категориальные данные в тип данных `category`. Зачастую это делается исключительно для оптимизации работы с большими данными.

- Если набор данных занимает значительный процент используемой оперативной памяти, рассмотрите возможность использования типа `category`.

- Если у вас очень серьёзные проблемы с производительностью, обратите внимание на использование типа `category`.

- Если вы решили использовать тип `category`, будьте осторожны при добавлении новой информации в вашу таблицу. Убедитесь, что вы собрали всю необходимую информацию, произведите предобработку данных и только после этого используйте преобразование типов.


## 6. Практика по Pandas <a class="anchor" id="6"></a>
[К содержанию](#0)

Датасет представляет собой таблицу с информацией о 300 тысячах поездок за первые пять дней сентября 2018 года и включает в себя следующую информацию:
- starttime — время начала поездки (дата, время);
- stoptime — время окончания поездки (дата, время);
- start station id — идентификатор стартовой стоянки;
- start station name — название стартовой стоянки;
- start station latitude, start station longitude — географическая широта и долгота стартовой стоянки;
- end station id — идентификатор конечной стоянки;
- end station name — название конечной стоянки;
- end station latitude, end station longitude — географическая широта и долгота конечной стоянки;
- bikeid — идентификатор велосипеда;
- usertype — тип пользователя (Customer — клиент с подпиской на 24 часа или на три дня,- Subscriber — подписчик с годовой арендой велосипеда);
- birth year — год рождения клиента;
- gender — пол клиента (0 — неизвестный, 1 — мужчина, 2 — женщина).

In [11]:
url = 'C:/Личные документы/Python/Py/data'
citibike_data = pd.read_csv(url+'/citibike-tripdata.csv', sep=',')
citibike_df = citibike_data.copy()
citibike_df['starttime'] = pd.to_datetime(citibike_df['starttime'], dayfirst=True)
citibike_df['stoptime'] = pd.to_datetime(citibike_df['stoptime'], dayfirst=True)


In [37]:
# количество не пустых значений в столбце 'start station id'
citibike_df['start station id'].count()
# количество строк
len(citibike_df['start station id'])
non_null_start_station_id = len(citibike_df['start station id']) - citibike_df['start station id'].count()
non_null_start_station_id

169

In [38]:
# идентификатор самой популярной стартовой стоянки.
citibike_df['start station id'].value_counts().nlargest(1)

281.0    1928
Name: start station id, dtype: int64

In [39]:
# самый популярный велосипед
citibike_df['bikeid'].value_counts()

33887    107
34358    104
34159    103
34258    103
33916    102
        ... 
18587      1
17324      1
29593      1
33157      1
25794      1
Name: bikeid, Length: 10004, dtype: int64

In [40]:
# Какой тип клиентов (столбец usertype) является преобладающим — Subscriber или Customer? 
# запишите долю клиентов преобладающего типа среди общего количества клиентов. 
# Ответ округлите до сотых.
subscribers = citibike_df['usertype'][citibike_df['usertype'] == 'Subscriber'].count()
customers = citibike_df['usertype'][citibike_df['usertype'] == 'Customer'].count()
print('Клиентов с подпиской:', subscribers)
print('Клиентов без подписки:', customers)
print('Доля клиентов с подпиской:', subscribers / len(citibike_df['usertype']))
print('Доля клиентов без подпиской:', customers / len(citibike_df['usertype']))
citibike_df['usertype'].value_counts(normalize=True)

Клиентов с подпиской: 232202
Клиентов без подписки: 67798
Доля клиентов с подпиской: 0.7740066666666666
Доля клиентов без подпиской: 0.22599333333333332


Subscriber    0.774007
Customer      0.225993
Name: usertype, dtype: float64

In [41]:
# Кто больше занимается велоспортом — мужчины или женщины? 
# В ответ запишите число поездок для той группы, у которой их больше.
male_count = citibike_df[citibike_df['gender']==1].shape[0]
female_count = citibike_df[citibike_df['gender']==2].shape[0]
print(max([male_count, female_count]))

183582


In [42]:
# Число уникальных стартовых и конечных стоянок, 
# которыми воспользовались клиенты, не равны друг другу
display(citibike_df['start station id'].nunique() != citibike_df['end station id'].nunique())

True

In [43]:
# В рассматриваемые дни минимальный возраст клиента составлял 10 лет
# добавим строку возраст
citibike_df['Age'] = citibike_df['birth year'].apply(lambda x: 2022 - x)
display(citibike_df['Age'].min() == 10 )
# или не добавляя строку возраста
citibike_df['birth year'].max() == 2012

False

False

In [44]:
# Самой непопулярной стартовой стоянкой из тех, которыми воспользовались клиенты, 
# является стоянка с названием "Eastern Pkwy & Washington Ave"
citibike_df['start station name'].value_counts()

Grand Army Plaza & Central Park S    1928
Central Park S & 6 Ave               1909
Pershing Square North                1873
12 Ave & W 40 St                     1845
West St & Chambers St                1835
                                     ... 
NYCBS Depot - GOW                      15
Franklin Ave & Empire Blvd             11
Railroad Ave & Kay Ave                  9
47 Ave & Skillman Ave                   7
Eastern Pkwy & Washington Ave           6
Name: start station name, Length: 759, dtype: int64

In [45]:
# Наибольшее количество поездок завершается на стоянке под названием "Liberty Light Rail"
citibike_df['end station name'].value_counts()

West St & Chambers St          1963
12 Ave & W 40 St               1909
Pershing Square North          1850
Central Park S & 6 Ave         1828
E 17 St & Broadway             1804
                               ... 
Exchange Place                    3
Union St                          1
Warren St                         1
Montrose Ave & Bushwick Ave       1
Liberty Light Rail                1
Name: end station name, Length: 765, dtype: int64

In [46]:
# Удаим признаки идентификаторов стоянок
citibike_df.info()
citibike_df = citibike_df.drop(['start station id', 'end station id'], axis=1)
citibike_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300000 entries, 0 to 299999
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   starttime                300000 non-null  datetime64[ns]
 1   stoptime                 300000 non-null  datetime64[ns]
 2   start station id         299831 non-null  float64       
 3   start station name       299831 non-null  object        
 4   start station latitude   300000 non-null  float64       
 5   start station longitude  300000 non-null  float64       
 6   end station id           299831 non-null  float64       
 7   end station name         299831 non-null  object        
 8   end station latitude     300000 non-null  float64       
 9   end station longitude    300000 non-null  float64       
 10  bikeid                   300000 non-null  int64         
 11  usertype                 300000 non-null  object        
 12  birth year      

In [47]:
# Замените признак birth year на более понятный признак возраста клиента age. 
# Годом отсчёта возраста выберите 2018 год. Столбец birth year удалите из таблицы
citibike_df['Age'] = citibike_df['birth year'].apply(lambda x: 2018 - x)
citibike_df = citibike_df.drop(['birth year'],axis=1)

In [48]:
# 6.9 сколько поездок совершили люди старше 60
citibike_df[citibike_df['Age'] > 60]

starttime                stoptime  \
53     2018-09-01 00:04:20.920 2018-09-01 00:11:21.618   
373    2018-09-01 00:25:47.674 2018-09-01 00:39:31.716   
405    2018-09-01 00:29:10.163 2018-09-01 00:44:08.611   
435    2018-09-01 00:31:21.725 2018-09-01 00:41:57.920   
457    2018-09-01 00:33:19.808 2018-09-01 00:37:29.927   
...                        ...                     ...   
299882 2018-09-05 19:07:25.401 2018-09-05 19:35:36.500   
299883 2018-09-05 19:07:24.913 2018-09-05 19:20:20.276   
299946 2018-09-05 19:08:06.462 2018-09-05 19:20:16.018   
299957 2018-09-05 19:08:12.577 2018-09-05 19:42:24.034   
299990 2018-09-05 19:08:26.264 2018-09-05 19:44:53.780   

                       start station name  start station latitude  \
53      Grand Army Plaza & Central Park S               40.764397   
373                   Mott St & Prince St               40.723180   
405                     Avenue D & E 3 St               40.720828   
435                    Broadway & W 56 St               40.765265   
457                  Warren St & Court St               40.686371   
...                                   ...                     ...   
299882            Amsterdam Ave & W 66 St               40.774667   
299883                   W 4 St & 7 Ave S               40.734011   
299946                Henry St & Grand St               40.714215   
299957              Market St & Cherry St               40.710762   
299990             Columbus Ave & W 72 St               40.777057   

        start station longitude           end station name  \
53                   -73.973715         Broadway & W 53 St   
373                  -73.994800        Mott St & Prince St   
405                  -73.977932         Avenue D & E 12 St   
435                  -73.981923           W 37 St & 10 Ave   
457                  -73.993833        Bond St & Bergen St   
...                         ...                        ...   
299882               -73.984706  Cathedral Pkwy & Broadway   
299883               -74.002939          E 2 St & Avenue A   
299946               -73.981346      Lafayette St & E 8 St   
299957               -73.994004   Catherine St & Monroe St   
299990               -73.978985       Hudson St & Reade St   

        end station latitude  end station longitude  bikeid    usertype  \
53                 40.763441             -73.982681   24824  Subscriber   
373                40.723180             -73.994800   26083  Subscriber   
405                40.725806             -73.974225   33436  Subscriber   
435                40.756604             -73.997901   19894  Subscriber   
457                40.684967             -73.986208   34173  Subscriber   
...                      ...                    ...     ...         ...   
299882             40.804213             -73.966991   16376  Subscriber   
299883             40.723077             -73.985836   29943  Subscriber   
299946             40.730207             -73.991026   30951  Subscriber   
299957             40.711174             -73.996826   31453  Subscriber   
299990             40.716250             -74.009106   26377  Subscriber   

        gender  Age  
53           2   63  
373          2   68  
405          2   61  
435          2   63  
457          1  118  
...        ...  ...  
299882       1   64  
299883       2   70  
299946       1   71  
299957       1   76  
299990       1   66  

[11837 rows x 12 columns]

In [50]:
# 6.10 Создайте признак длительности поездки trip duration
citibike_df['trip duration'] = citibike_df['stoptime'] - citibike_df['starttime']
# запишем среднюю продолжительность поездки в переменную mean_trip
mean_trip = citibike_df['trip duration'].mean()
# вычислим количество секунд (округлив до целого)
round(mean_trip.total_seconds(),2)

1126.56

In [51]:
# 6.11 Создайте «признак-мигалку» weekend, который равен 1, 
# если поездка начиналась в выходной день (суббота или воскресенье), 
# и 0 — в противном случае. Выясните, сколько поездок начиналось в выходные.

# запишем в переменную день недели
day_of_week = citibike_df['starttime'].dt.dayofweek
# определим функцию, которая вернёт 1 в выходные (5,6) и 0 в будни (0,1,2,3,4)
def is_weekend(arg):
    if arg == 5 or arg ==6:
        return 1
    else:
        return 0
# добавим столбец применив функцию к переменной    
citibike_df['weekend'] = day_of_week.apply(is_weekend)
citibike_df[citibike_df['weekend'] == 1]

starttime                stoptime  \
0      2018-09-01 00:00:05.269 2018-09-01 00:27:20.634   
1      2018-09-01 00:00:11.281 2018-09-01 00:02:23.481   
2      2018-09-01 00:00:20.649 2018-09-01 00:55:58.547   
3      2018-09-01 00:00:21.746 2018-09-01 00:07:38.583   
4      2018-09-01 00:00:27.315 2018-09-01 02:21:25.308   
...                        ...                     ...   
248507 2018-09-01 19:01:39.660 2018-09-01 19:09:34.516   
248516 2018-09-01 21:53:55.244 2018-09-01 21:59:19.266   
248520 2018-09-02 09:46:26.247 2018-09-02 09:51:56.662   
248525 2018-09-02 09:55:13.467 2018-09-02 10:01:29.366   
248549 2018-09-02 16:18:31.305 2018-09-02 16:36:00.042   

                     start station name  start station latitude  \
0          MacDougal St & Washington Sq               40.732264   
1       Cadman Plaza West & Montague St               40.693830   
2                       1 Ave & E 62 St               40.761227   
3               St James Pl & Oliver St               40.713079   
4                       W 13 St & 6 Ave               40.736494   
...                                 ...                     ...   
248507           Concord St & Bridge St               40.697666   
248516           Concord St & Bridge St               40.697666   
248520           Concord St & Bridge St               40.697666   
248525           Concord St & Bridge St               40.697666   
248549           Concord St & Bridge St               40.697666   

        start station longitude                 end station name  \
0                    -73.998522         Clinton Ave & Myrtle Ave   
1                    -73.990539       Schermerhorn St & Court St   
2                    -73.960940                  Smith St & 3 St   
3                    -73.998512              Park Pl & Church St   
4                    -73.997044                 W 4 St & 7 Ave S   
...                         ...                              ...   
248507               -73.984764         Front St & Washington St   
248516               -73.984764         Front St & Washington St   
248520               -73.984764          Pearl St & Anchorage Pl   
248525               -73.984764  Cadman Plaza West & Montague St   
248549               -73.984764                     4 Ave & 9 St   

        end station latitude  end station longitude  bikeid    usertype  \
0                  40.693261             -73.968896   25577  Subscriber   
1                  40.691029             -73.991834   34377  Subscriber   
2                  40.678724             -73.995991   30496  Subscriber   
3                  40.713342             -74.009355   28866  Subscriber   
4                  40.734011             -74.002939   20943    Customer   
...                      ...                    ...     ...         ...   
248507             40.702551             -73.989402   27688  Subscriber   
248516             40.702551             -73.989402   14909  Subscriber   
248520             40.702819             -73.987658   25959  Subscriber   
248525             40.693830             -73.990539   16371  Subscriber   
248549             40.670513             -73.988766   21282  Subscriber   

        gender  Age          trip duration  weekend  
0            1   38 0 days 00:27:15.365000        1  
1            0   49 0 days 00:02:12.200000        1  
2            1   43 0 days 00:55:37.898000        1  
3            2   34 0 days 00:07:16.837000        1  
4            1   24 0 days 02:20:57.993000        1  
...        ...  ...                    ...      ...  
248507       1   31 0 days 00:07:54.856000        1  
248516       1   32 0 days 00:05:24.022000        1  
248520       1   31 0 days 00:05:30.415000        1  
248525       1   32 0 days 00:06:15.899000        1  
248549       1   31 0 days 00:17:28.737000        1  

[115135 rows x 14 columns]

In [52]:
# 6.12 Создайте признак времени суток поездки time_of_day
time_of_day = citibike_df['starttime'].dt.hour

def day_time(arg):
    if 0 <= arg <= 6:
        return 'night'
    elif 6 < arg <= 12:
        return 'morning'
    elif 12 < arg <= 18:
        return 'day'
    elif 18 < arg <= 24:
        return 'evening'

citibike_df['time of day'] = time_of_day.apply(day_time)
day_trip = citibike_df[citibike_df['time of day']=='day'].count()
night_trip = citibike_df[citibike_df['time of day']=='night'].count()
round(day_trip / night_trip)

starttime                  9.0
stoptime                   9.0
start station name         9.0
start station latitude     9.0
start station longitude    9.0
end station name           9.0
end station latitude       9.0
end station longitude      9.0
bikeid                     9.0
usertype                   9.0
gender                     9.0
Age                        9.0
trip duration              9.0
weekend                    9.0
time of day                9.0
dtype: float64

[К содержанию](#0)